cpp/src/strings/copying/shift.cu

```diff
@@ -0,0 +1,149 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/strings/detail/copying.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+struct adjust_offsets_fn {
+  column_device_view const d_column;
+  string_view const d_filler;
+  size_type const offset;
+
+  __device__ offset_type operator()(size_type idx)
+  {
+    if (offset < 0) {
+      auto const first      = d_column.element<offset_type>(-offset);
+      auto const last_index = d_column.size() + offset;
+      if (idx < last_index) {
+        return d_column.element<offset_type>(idx - offset) - first;
+      } else {
+        auto const last = d_column.element<offset_type>(d_column.size() - 1);
+        return (last - first) + ((idx - last_index + 1) * d_filler.size_bytes());
+      }
+    } else {
+      if (idx < offset) {
+        return idx * d_filler.size_bytes();
+      } else {
+        auto const total_filler = d_filler.size_bytes() * offset;
+        return total_filler + d_column.element<offset_type>(idx - offset);
+      }
+    }
+  }
+};
+
+struct shift_chars_fn {
+  column_device_view const d_column;
+  string_view const d_filler;
+  size_type const offset;
+
+  __device__ char operator()(size_type idx)
+  {
+    if (offset < 0) {
+      auto const last_index = -offset;
+      if (idx < last_index) {
+        auto const first_index = d_column.size() + offset;
+        return d_column.element<char>(idx + first_index);
+      } else {
+        auto const char_index = idx - last_index;
+        return d_filler.data()[char_index % d_filler.size_bytes()];
+      }
+    } else {
+      if (idx < offset) {
+        return d_filler.data()[idx % d_filler.size_bytes()];
+      } else {
+        return d_column.element<char>(idx - offset);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> shift(strings_column_view const& input,
+                              size_type offset,
+                              scalar const& fill_value,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
+{
+  auto d_fill_str = dynamic_cast<string_scalar const&>(fill_value).value(stream);
```

Comment:
This is a downcast of `scalar` to `string_scalar`. Since `scalar` could technically be another scalar-type at runtime, a `static_cast` here will not be able to guarantee the downcast is correct since `static_cast` is a compile time check. The `dynamic_cast` provides a run-time check to ensure the cast is allowed.

Reply:
should we replace `static_cast<*_scalar&>` with `dynamic_cast` at many places in our code?

Reply:
Only if they are wrong :-). I see many examples that use `static_cast` to a scalar type that has been resolved directly by the `type-dispatcher`. In these cases, the _compiler_ has already guaranteed the type and the downcast via `static_cast` is ok.

The `type-dispatcher` used by `cudf::detail::shift` is done on the `column_view` type and not the `scalar's` type. There is a check at the beginning of the API to make sure the type of the `column_view` and `scalar` match though this is a run-time check. So I guess this run-time check will technically allow a `static_cast` here.